src/mlpack/methods/ann/activation_functions/tanh_exponential_function.hpp

```diff
@@ -0,0 +1,83 @@
+/**
+ * @file methods/ann/activation_functions/tanh_exponential_function.hpp
+ * @author Mayank Raj
+ *
+ * Definition and implementation of the Tanh exponential  function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef TANH_EXPONENTIAL_FUNCTION_HPP_INCLUDED
+#define TANH_EXPONENTIAL_FUNCTION_HPP_INCLUDED
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The TanhExp function, defined by
+ *
+ * @f{eqnarray*}{
+ * f(x) = x * tanh(e^x)\\
+ * f'(x) = tanh(e^x) - x*e^x*(tanh(e^x)^2 - 1)\\
+ * @f}
+ */
+ class TanhExpFunction
+{
+ public:
+  /**
+   * Computes the TanhExp function.
+   *
+   * @param x Input data.
+   * @return f(x).
+   */
+  static double Fn(const double x)
+  {
+    return x*std::tanh(std::exp(x));
+  }
+
+  /**
+   * Computes the TanhExp function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, OutputVecType& y)
+  {
+    y = x*arma::tanh(arma::exp(x));
+  }
+
+  /**
+   * Computes the first derivative of the TanhExp function.
+   *
+   * @param y Input activation.
+   * @return f'(x)
+   */
+  static double Deriv(const double y)
+  {
+    return std::tanh(std::exp(y)) -
+    y*std::exp(y)*(std::pow(std::tanh(std::exp(y)),2) - 1);
+  }
+
+  /**
+   * Computes the first derivatives of the tanh function.
+   *
+   * @param y Input activations.
+   * @param x The resulting derivatives.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Deriv(const InputVecType& y, OutputVecType& x)
+  {
+    x = arma::tanh(arma::exp(y)) -
```

Comment:
yes you are right, i missed the transformation.

Reply:
wait no it's hyperbolic so 1 - tanh^2(x) = sech^2(x).

Reply:
Yes, I used `tanh = (e^x - e^(-x)) / ((e^x + e^(-x))` mainly because there is no `sech` in armadillo.

Reply:
`(e^x * x)/(cosh^2(e^x)) + (sinh(e^x)) / (cosh(e^x))` = `(e^x * x)*(sech^2(e^x)) + (tanhh(e^x)) `
                                                                                   = `(e^x * x)*(1 - tanh^2(x) ) + (tanhh(e^x)) `
                                                                                   = `- (e^x * x)*( tanh^2(x) - 1 ) + (tanhh(e^x)) `

Reply:
So looks like the the formula is still the same, do you agree that what I posted above is correct `(e^x * x)/(cosh^2(e^x)) + (sinh(e^x)) / (cosh(e^x))`? If so, can you show me the transformation for what I posted to what you implemented, because the transformation `(e^x * x)/(cosh^2(e^x)) + (sinh(e^x)) / (cosh(e^x)) = - (e^x * x)*( tanh^2(x) - 1 ) + (tanhh(e^x))` you posted above isn't correct maybe there was a copy/paste error? We can also check if another frameworks implements the function and check against those values, just to be sure.

Reply:
![Screenshot (56)](https://user-images.githubusercontent.com/54908605/111957549-29afee80-8b12-11eb-87ba-d523a124e68a.png)
The values are matching for both the expression.

Reply:
![nmnnm](https://user-images.githubusercontent.com/54908605/111959599-af349e00-8b14-11eb-9264-84bd835c99af.jpeg)
The transformation which I used.